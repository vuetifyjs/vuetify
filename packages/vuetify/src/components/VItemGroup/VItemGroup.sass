@use '../../styles/tools'
@use './variables' as *

@include tools.layer('components')
  .v-item-group
    flex: 0 1 auto
    max-width: 100%
    position: relative
    transition: $item-group-transition
