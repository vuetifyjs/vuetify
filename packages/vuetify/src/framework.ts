// Composables
import { createDate, DateAdapterSymbol, DateOptionsSymbol } from '@/composables/date/date'
import { createDefaults, DefaultsSymbol } from '@/composables/defaults'
import { createDisplay, DisplaySymbol } from '@/composables/display'
import { createGoTo, GoToSymbol } from '@/composables/goto'
import { createIcons, IconSymbol } from '@/composables/icons'
import { createLocale, LocaleSymbol } from '@/composables/locale'
import { createTheme, ThemeSymbol } from '@/composables/theme'

// Utilities
import { nextTick, reactive } from 'vue'
import { defineComponent, getUid, IN_BROWSER, mergeDeep } from '@/util'

// Types
import type { App, ComponentPublicInstance, InjectionKey } from 'vue'
import type { DateOptions } from '@/composables/date'
import type { DefaultsOptions } from '@/composables/defaults'
import type { DisplayOptions, SSROptions } from '@/composables/display'
import type { GoToOptions } from '@/composables/goto'
import type { IconOptions } from '@/composables/icons'
import type { LocaleOptions, RtlOptions } from '@/composables/locale'
import type { ThemeOptions } from '@/composables/theme'
export * from './composables'
export type { DateOptions, DateInstance, DateModule } from '@/composables/date'

export interface VuetifyOptions {
  aliases?: Record<string, any>
  blueprint?: Blueprint
  components?: Record<string, any>
  date?: DateOptions
  directives?: Record<string, any>
  defaults?: DefaultsOptions
  display?: DisplayOptions
  goTo?: GoToOptions
  theme?: ThemeOptions
  icons?: IconOptions
  locale?: LocaleOptions & RtlOptions
  ssr?: SSROptions
}

export interface Blueprint extends Omit<VuetifyOptions, 'blueprint'> {}

export function createVuetify (vuetify: VuetifyOptions = {}) {
  const { blueprint, ...rest } = vuetify
  const options: VuetifyOptions = mergeDeep(blueprint, rest)
  const {
    aliases = {},
    components = {},
    directives = {},
  } = options

  const defaults = createDefaults(options.defaults)
  const display = createDisplay(options.display, options.ssr)
  const theme = createTheme(options.theme)
  const icons = createIcons(options.icons)
  const locale = createLocale(options.locale)
  const date = createDate(options.date, locale)
  const goTo = createGoTo(options.goTo, locale)

  const install = (app: App) => {
    for (const key in directives) {
      app.directive(key, directives[key])
    }

    for (const key in components) {
      app.component(key, components[key])
    }

    for (const key in aliases) {
      app.component(key, defineComponent({
        ...aliases[key],
        name: key,
        aliasName: aliases[key].name,
      }))
    }

    theme.install(app)

    app.provide(DefaultsSymbol, defaults)
    app.provide(DisplaySymbol, display)
    app.provide(ThemeSymbol, theme)
    app.provide(IconSymbol, icons)
    app.provide(LocaleSymbol, locale)
    app.provide(DateOptionsSymbol, date.options)
    app.provide(DateAdapterSymbol, date.instance)
    app.provide(GoToSymbol, goTo)

    if (IN_BROWSER && options.ssr) {
      if (app.$nuxt) {
        app.$nuxt.hook('app:suspense:resolve', () => {
          display.update()
        })
      } else {
        const { mount } = app
        app.mount = (...args) => {
          const vm = mount(...args)
          nextTick(() => display.update())
          app.mount = mount
          return vm
        }
      }
    }

    getUid.reset()

    if (typeof __VUE_OPTIONS_API__ !== 'boolean' || __VUE_OPTIONS_API__) {
      app.mixin({
        computed: {
          $vuetify () {
            return reactive({
              defaults: inject.call(this, DefaultsSymbol),
              display: inject.call(this, DisplaySymbol),
              theme: inject.call(this, ThemeSymbol),
              icons: inject.call(this, IconSymbol),
              locale: inject.call(this, LocaleSymbol),
              date: inject.call(this, DateAdapterSymbol),
            })
          },
        },
      })
    }
  }

  return {
    install,
    defaults,
    display,
    theme,
    icons,
    locale,
    date,
    goTo,
  }
}

export const version = __VUETIFY_VERSION__
createVuetify.version = version

// Vue's inject() can only be used in setup
function inject (this: ComponentPublicInstance, key: InjectionKey<any> | string) {
  const vm = this.$

  const provides = vm.parent?.provides ?? vm.vnode.appContext?.provides

  if (provides && (key as any) in provides) {
    return provides[(key as string)]
  }
}
