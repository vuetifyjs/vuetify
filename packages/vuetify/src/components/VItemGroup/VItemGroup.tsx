// Styles
import './VItemGroup.sass'

// Composables
import { makeComponentProps } from '@/composables/component'
import { makeGroupProps, useGroup } from '@/composables/group'
import { makeTagProps } from '@/composables/tag'
import { makeThemeProps, provideTheme } from '@/composables/theme'

// Utilities
import { genericComponent, propsFactory } from '@/util'

// Types
import type { GenericProps } from '@/util'

export const VItemGroupSymbol = Symbol.for('vuetify:v-item-group')

export const makeVItemGroupProps = propsFactory({
  ...makeComponentProps(),
  ...makeGroupProps({
    selectedClass: 'v-item--selected',
  }),
  ...makeTagProps(),
  ...makeThemeProps(),
}, 'VItemGroup')

type VItemGroupSlots = {
  default: {
    isSelected: (id: string) => boolean
    select: (id: string, value: boolean) => void
    next: () => void
    prev: () => void
    selected: readonly string[]
  }
}

export const VItemGroup = genericComponent<new <T>(
  props: {
    modelValue?: T
    'onUpdate:modelValue'?: (value: T) => void
  },
  slots: VItemGroupSlots,
) => GenericProps<typeof props, typeof slots>>()({
  name: 'VItemGroup',

  props: makeVItemGroupProps(),

  emits: {
    'update:modelValue': (value: any) => true,
  },

  setup (props, { slots }) {
    const { themeClasses } = provideTheme(props)
    const { isSelected, select, next, prev, selected } = useGroup(props, VItemGroupSymbol)

    return () => (
      <props.tag
        class={[
          'v-item-group',
          themeClasses.value,
          props.class,
        ]}
        style={ props.style }
      >
        { slots.default?.({
          isSelected,
          select,
          next,
          prev,
          selected: selected.value,
        })}
      </props.tag>
    )
  },
})

export type VItemGroup = InstanceType<typeof VItemGroup>
