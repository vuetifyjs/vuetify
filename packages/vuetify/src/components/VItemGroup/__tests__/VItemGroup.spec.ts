// Components
import { VItem } from '../VItem'
import { VItemGroup } from '../VItemGroup'

// Utilities
import { mount } from '@vue/test-utils'
import { h } from 'vue'
import { createVuetify } from '@/framework'

describe('VItemGroup', () => {
  const vuetify = createVuetify()
  const mountFunction = (options = {}) => {
    return mount(VItemGroup, {
      ...options,
      global: {
        plugins: [vuetify],
      },
    })
  }

  const defaultSlot = () => [
    h(VItem, { value: 'foo' }, {
      default: ({ toggle }: any) => h('div', { id: 'item', onClick: toggle }, ['foo']),
    }),
    h(VItem, { value: 'bar' }, {
      default: ({ toggle }: any) => h('div', { id: 'item', onClick: toggle }, ['bar']),
    }),
  ]

  it('should update state from child clicks', async () => {
    const wrapper = mountFunction({
      slots: {
        default: defaultSlot,
      },
    })

    const items = wrapper.findAll('#item')

    await items[0].trigger('click')
    await items[1].trigger('click')

    const events = wrapper.emitted('update:modelValue')

    expect(events).toEqual([['foo'], ['bar']])
  })

  it('should not deselect value when using mandatory prop', async () => {
    const wrapper = mountFunction({
      props: {
        mandatory: true,
        modelValue: 'foo',
      },
      slots: {
        default: defaultSlot,
      },
    })

    const items = wrapper.findAll('#item')

    await items[0].trigger('click')

    const events = wrapper.emitted('update:modelValue')

    expect(events).toBeUndefined()
  })

  it('should update a multiple item group', async () => {
    const wrapper = mountFunction({
      props: {
        multiple: true,
        modelValue: ['foo'],
      },
      slots: {
        default: defaultSlot,
      },
    })

    const items = wrapper.findAll('#item')

    await items[0].trigger('click')
    await items[1].trigger('click')

    const events = wrapper.emitted('update:modelValue')

    expect(events).toEqual([
      [[]],
      [['bar']],
    ])
  })

  it('should ignore disabled items', async () => {
    const wrapper = mountFunction({
      props: {
        multiple: true,
      },
      slots: {
        default: () => [
          h(VItem, { value: 'foo', disabled: true }, {
            default: ({ toggle }: any) => h('div', { id: 'item', onClick: toggle }, ['foo']),
          }),
          h(VItem, { value: 'bar' }, {
            default: ({ toggle }: any) => h('div', { id: 'item', onClick: toggle }, ['bar']),
          }),
        ],
      },
    })

    const items = wrapper.findAll('#item')

    await items[0].trigger('click')
    await items[1].trigger('click')

    const events = wrapper.emitted('update:modelValue')

    expect(events).toEqual([
      [['bar']],
    ])
  })

  // https://github.com/vuetifyjs/vuetify/issues/5384
  it('should not unregister children when is destroyed', async () => {
    const change = vi.fn()
    const wrapper = mountFunction({
      props: {
        modelValue: 'foo',
        'onUpdate:modelValue': change,
      },
      slots: {
        default: () => h(VItem, { value: 'foo' }, () => ['foo']),
      },
    })

    wrapper.unmount()

    expect(change).not.toHaveBeenCalled()
  })

  it('should render with a specified tag when the tag prop is provided with a value', () => {
    const wrapper = mountFunction({
      props: {
        tag: 'button',
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
