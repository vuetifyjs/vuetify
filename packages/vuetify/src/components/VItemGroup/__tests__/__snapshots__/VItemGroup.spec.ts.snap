// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VItemGroup > should render with a specified tag when the tag prop is provided with a value 1`] = `
"<button class="v-item-group v-theme--light">
  <!---->
</button>"
`;
